XenModel/Actions/AsyncAction.cs

```diff
@@ -268,7 +268,107 @@ public void DestroyTask()
+                    //then log every 30sec
+                    int currDebug = (int)(DateTime.Now - startTime).TotalSeconds / 30;
+                    if (currDebug > lastDebug)
+                    {
+                        lastDebug = currDebug;
+                        log.DebugFormat("Polling for action {0}", Description);
+                    }
+
+                    if (Poll(start, finish))
+                        break;
+                    Thread.Sleep(900);
+                }
+            }
+            finally
+            {
+                DestroyTask();
+            }
+        }
+
+        /// <summary>
+        /// Polls task and returns whether it is completed
+        /// </summary>
+        private bool Poll(double start, double finish)
+        {
+            Session session = Session;
+            Task task;
+            Result = "";
+
+            try
+            {
+                task = (Task)DoWithSessionRetry(ref session, (Task.TaskGetRecordOp)Task.get_record,
+                    RelatedTask.opaque_ref);
+            }
+            catch (Failure exn)
+            {
+                if (exn.ErrorDescription.Count > 1 && 
+                    exn.ErrorDescription[0] == Failure.HANDLE_INVALID &&
+                    exn.ErrorDescription[1] == "task")
+                {
+                    log.WarnFormat("Invalid task handle {0} - task is finished.", RelatedTask.opaque_ref);
+                    PercentComplete = (int)finish;
+                    return true;
+                }
+                else
+                    throw;
+            }
+            finally
+            {
+                Session = session;
+            }
+
+            Tick((int)(start + task.progress * (finish - start)),
+                task.Description() == "" ? Description : task.Description());
+
+            switch (task.status)
+            {
+                case task_status_type.failure:
+                    if (task.error_info.Length > 1 &&
+                        task.error_info[0] == Failure.HANDLE_INVALID &&
+                        task.error_info[1] == "task")
+                    {
+                        log.WarnFormat("Invalid task handle {0} - task is finished.", RelatedTask.opaque_ref);
+                        PercentComplete = (int)finish;
+                        return true;
+                    }
+                    else
+                    {
+                        log.WarnFormat("Task {0} failed: {1}", RelatedTask.opaque_ref,
+                            task.error_info.Length > 0 ? task.error_info[0] : "Unknown failure");
+                        throw new Failure(task.error_info);
+                    }
+
+                case task_status_type.success:
+                    log.InfoFormat("Task {0} finished successfully", RelatedTask.opaque_ref);
+                    if (task.result != "") // Work around CA-6597
+                    {
+                        Match m = Regex.Match(Result, "<value>(.*)</value>");

```

Comment:
It's not ideal to parse XML with RegEx.
However I note that this is from the existing code in `TaskPoller.cs`, so it is probably safer to leave it as is.

Reply:
I agree; it can be revisited at a later time.